QIR oracle generation sample

msoeken opened the pull request:
Implements #496

cgranade commented on samples/qir/oracle-generator/qsharp/Program.qs:
```diff
@@ -0,0 +1,50 @@
+        }
+    }
+
+    internal function InitOracleCompilerFor<'In, 'Out>(func : 'In -> 'Out) : Unit {
```
What is this function used for?

msoeken replied:
I'll add a comment. But for now it's actually only used as a workaround to avoid that all the operations and functions referred to in this function are included into the generated QIR code. If they are not used, they are going to be stripped.

cgranade replied:
Makes sense, I figured it was something like that. As you say, adding a comment can help the reader understand why that workaround is needed.

cgranade commented on samples/qir/oracle-generator/qsharp/Program.qs:
```diff
@@ -0,0 +1,50 @@
+}
+
+namespace OracleCompiler.Classical {
+    internal function Majority3(a : Bool, b : Bool, c : Bool) : Bool {
```
Some comments here could help as well.

cgranade commented on samples/qir/oracle-generator/qsharp/project.csproj:
```diff
@@ -0,0 +1,14 @@
+<Project Sdk="Microsoft.Quantum.Sdk/0.16.2105142209-alpha">
```
Does this require an alpha version? Is it possible to use a released version?

msoeken replied:
I am not sure, I'll try to use the version of the upcoming release.

msoeken replied:
The package `Microsoft.Quantum.Qir.Runtime` does not exist in a released version yet. I suggest to wait until the May release and then change the version number, and keep this alpha version for now.

cgranade replied:
Makes sense; just wanted to double-check on that. I think it's fine to keep the version number as-is, then, but I would emphasize in the README that QIR generation is still in preview to help set user expectations.

cgranade reviewed (commented):
This review is preliminary, given the size of your PR, but wanted to give early feedback. Thanks!

cgranade commented:
It looks like the build failure is because the agent tests against .NET Core SDK 3.1 only, matching our end-user installation instructions. We should discuss how to generalize the build for this repo since this sample requires 5.0.

msoeken commented on samples/qir/oracle-generator/oracle-compiler/oracle_compiler.cpp:
```diff
@@ -0,0 +1,144 @@
+#include <iostream>
+#include <unordered_map>
+
+#include <fmt/format.h>
```
I totally agree. I will spend some time to add documentation to all these files.

msoeken commented on samples/qir/oracle-generator/oracle-compiler/oracle_compiler.cpp:
```diff
@@ -0,0 +1,144 @@
+    /* collapse and resynthesize if number of inputs is small */
+    if (xag.num_pis() <= 8)
+    {
+        auto klut = single_node_network<klut_network>(xag);
```
Exactly, I'll also add comments here.

msoeken commented:
> It looks like the build failure is because the agent tests against .NET Core SDK 3.1 only, matching our end-user installation instructions. We should discuss how to generalize the build for this repo since this sample requires 5.0.

I am not even sure whether .NET 5 is required. I am using it as a default when setting up projects. I'll check whether it can work with 3.1 to avoid this error.

cgranade commented:
> > It looks like the build failure is because the agent tests against .NET Core SDK 3.1 only, matching our end-user installation instructions. We should discuss how to generalize the build for this repo since this sample requires 5.0.
> 
> I am not even sure whether .NET 5 is required. I am using it as a default when setting up projects. I'll check whether it can work with 3.1 to avoid this error.

Makes sense. At the moment, the Quantum Development Kit officially requires 3.1 to be installed, since that's the most up-to-date LTS release, such that if our samples can work for users following those setup instructions that would be ideal. Thanks for looking into this!

msoeken commented:
Thanks a lot for your comments @cgranade. I addressed all but one of your comments and also added more comments to the source files. The missing comment of yours concerns whether we can use a released version for this sample, but none of the released version suffice so far. Therefore, I suggest to address this comment once the next release is out.

msoeken commented:
@cgranade Thanks for all your comments. I have just recently updated the README to meet the expectations of the sample. Have all your comments been addressed and can the PR be approved?

cgranade commented on samples/interoperability/qrng/NuGet.config:
```diff
@@ -0,0 +1,7 @@
+<configuration>
+  <packageSources>
+    <add key="qdk-alpha" value="https://pkgs.dev.azure.com/ms-quantum-public/Microsoft Quantum (public)/_packaging/alpha/nuget/v3/index.json" protocolVersion="3" />
+    <add key="local" value="/home/mathias/code/nugets" />
```
It looks like this line may not have been intended for inclusion in the sample?

msoeken replied:
In fact, the whole file is not intended for inclusion. Might have added it by mistake.

cgranade commented on samples/qir/oracle-generator/oracle-generator/oracle_generator.cpp:
```diff
@@ -0,0 +1,203 @@
+#include <mockturtle/networks/xag.hpp>
+#include <mockturtle/utils/cost_functions.hpp>
+
+// W modify LLVM through the object model provided by the LLVM C++ API.
```
```suggestion
// We modify LLVM through the object model provided by the LLVM C++ API.
```

cgranade reviewed (approved):
LGTM! Just had a couple minor comments left.
